### DISCUSSION TITLE
Impersonation

### DISCUSSION THREAD

**Discussion by ElliotHYLee** [pocketbase/pocketbase#3401] [Q&A] [answered]
As an admin, is it possible to get a user's token?

I wish to make a function that an admin can view the pages as a user without knowing the user's password or so. The simplest way to do this is to get the user's token and do the rest..

Best regards.

**Comment by ganigeorgiev** [comment, marked as answer]
Not out of the box at the moment. It is in the roadmap and will be eventually implemented in the future as part of https://github.com/pocketbase/pocketbase/issues/718#issuecomment-1270025193.

For now you can register a custom admin route. Here is a minimal example with the JS `pb_hooks` (_I haven't tested it_):

```js
routerAdd("POST", "/impersonate", (c) => {
    const data = new DynamicModel({
        email: "",
    })

    // read the request into the data variable
    c.bind(data)

    // find the user associated with the email
    const user = $app.dao().findAuthRecordByEmail("users", data.email)

    // generate an auth token and return it together with the the user as response
    return $apis.recordAuthResponse($app, c, user, null);
}, $apis.requireAdminAuth())
```

And on the client side:

```js
import PocketBase, { LocalAuthStore } from "pocketbase";

// create an admin SDK instance
const adminClient = new PocketBase("http://127.0.0.1:8090")

await adminClient.admins.authWithPassword("admin@example.com", "1234567890")

const result = await adminClient.send("/impersonate", {
    method: "POST",
    body: { email: "user@example.com" },
});

// create new user SDK instance (with a different local storage key)
const userClient = new PocketBase(adminClient.baseUrl, new LocalAuthStore("pb_impersonate"));
userClient.authStore.save(result.token, result.record);

// do something with the userClient instance...
```

**Reply by ElliotHYLee** [reply]
I must say this is smooooth

**Reply by GuptaSiddhant** [reply]
Works well. Thanks!

**Comment by nayakanirudh** [comment]
Whoever is migrating from a much earlier version and find this script not working, here's the updated one working with 0.26.6
Even though there is a new impersonate API here https://github.com/pocketbase/pocketbase/issues/718#issuecomment-1270025193, I wanted to continue to use this hook :)

```js
routerAdd("POST", "/api/admins/impersonate", (c) => {
    const data = new DynamicModel({
        email: "",
    })

    // read the request into the data variable
    c.bindBody(data)

    // find the user associated with the email
    const user = $app.findAuthRecordByEmail("users", data.email)

    // generate an auth token and return it together with the the user as response
    return $apis.recordAuthResponse(c, user, null);
}, $apis.requireSuperuserAuth())
```

**Reply by ElliotHYLee** [reply]
Thanks. I found this helpful. For now, I've also decided to go with it because my frontend and extra backend cannot verify whether the impersonated token is valid or not. As of now, the only way is to refresh the token, which can't be done in this case.
